### ISSUE TITLE
How to use role in PDF

### ISSUE THREAD

**Issue by JBR69** [asciidoctor/asciidoctor-pdf#652]
Hi,

Is it possible to use a specific role in YAML ?

Exemple:
 my asciidoc 
`Type the user password (by default [defvalue]#AzerTy007# ),`

in my CSS I have this class
`.defvalue{background:#444;color:#00ffff;font-family:monospace;font-size:1em}`

and I write this in my YAML
```
defvalue: 
    background_color: 444444
    font_color: 00ffff
    font_family: monospace 
```

It's OK in HTML but not in PDF.
So, is it possible ? And if yes, what's wrong in my YAML ?

ty

**Comment by git-ere** [comment]
I came over the same problem some days ago, see my comment in [Issue #350 ](https://github.com/asciidoctor/asciidoctor-pdf/pull/350?_pjax=%23js-repo-pjax-container)

**Comment by JBR69** [comment]
thank you, I close this issue and i'll try your solution

**Comment by git-ere** [comment]
pls let me know, if it works for you and if you think this possible fits into the "picture", If yes, I can do a fork and a merge request

**Comment by JBR69** [comment]
Sorry I didn't use your solution because I adapted the code that I proposed in issue #476 which modified only the file transform.rb

Line 211
```
when :span
      # ****************************************************************************
      # Add code                
      if (!attrs[:class].nil? and attrs[:class] != '')          
        pvalue = attrs[:class]                      

        font_color = @theme[%(#{pvalue}_font_color)]            
        if (!font_color.nil? and font_color != '')
            if font_color.start_with? '#'
              # hexa color code
              font_color = font_color[1..-1] 
            end
            fragment[:color] = font_color            
        end

        font_family = @theme[%(#{pvalue}_font_family)]
        if (!font_family.nil? and font_family != '')
            fragment[:font] = font_family
        end  

        font_size = @theme[%(#{pvalue}_font_size)]
        if (!font_size.nil? and font_size != '')
            fragment[:size] = font_size
        end

        font_style = @theme[%(#{pvalue}_font_style)]
        if (!font_style.nil? and font_style != '')
            if font_style.include? 'italic'
              styles << :italic
            end
            if font_style.include? 'bold'            
              styles << :bold
            end
            if font_style.include? 'underline'
              styles << :underline
            end
            if font_style.include? 'subscript'
              styles << :subscript
            end
            if font_style.include? 'superscript'
              styles << :superscript
            end
            if font_style.include? 'strikethrough'
              styles << :strikethrough
            end          
        end                             
      end
      # ****************************************************************************
        
      # span logic with normal style parsing
      if (inline_styles = attrs[:style])
       .......

```
And in th YAML file
```
# Colors
aqua: 
    font_color: 00ffff
blue: 
    font_color: 0000ff
green: 
    font_color: 008000
orange: 
    font_color: ffa500
red: 
    font_color: ff0000
yellow: 
    font_color: ffff00

# Specific CSS classes
underline:
    font_style: underline
    
userinput:
    font_color: ff8c00  
    font_style: underline bold italic
    font_size: round($base_font_size * 1.05)

defvalue:
    font_color: 00ffff
    font_style: italic
    font_size: round($base_font_size * 0.95)
```

**Comment by namkhaib** [comment]
I'am new on AsciiDoc, AsciiDoctor and Ruby, what's the location of the transform.rb file? My english is bad, sorry ;)

**Comment by namkhaib** [comment]
I finally found the file (/usr/lib/ruby/gems/2.4.0/gems/asciidoctor-pdf-1.5.0.alpha.16/lib/asciidoctor-pdf/formatted_text, right?), but my file looks this:
(and <code>when :span</code> starts at line 196)

    when :span
      # NOTE spaces in style attribute value are superfluous, for our purpose; split drops record after trailing ;
      attrs[:style].tr(' ', '').split(';').each do |style|
        pname, pvalue = style.split(':', 2)
        case pname
        when 'color'
          # QUESTION should we check whether the value is a valid hex color?
          unless fragment[:color]
            case pvalue.length
            when 6
              fragment[:color] = pvalue
            when 7
              fragment[:color] = pvalue.slice(1, 6) if pvalue.start_with?('#')
            # QUESTION should we support the 3 character form?
            #when 3
            #  fragment[:color] = pvalue.each_char.map {|c| c * 2 }.join
            #when 4
            #  fragment[:color] = pvalue.slice(1, 3).each_char.map {|c| c * 2 }.join if pvalue.start_with?('#')
            end
          end
        when 'font-weight'
          if pvalue == 'bold'
            styles << :bold
          end
        when 'font-style'
          if pvalue == 'italic'
            styles << :italic
          end
        # TODO text-transform
        end
      end if attrs.key?(:style)
    end

**Comment by JBR69** [comment]
After the code you copied, you must have this.
(Add the code between # ----- and # -----)

```
# TODO we could limit to select tags, but doesn't seem to really affect performance
    attrs[:class].split.each do |class_name|
      case class_name
      when 'underline'
        styles << :underline
      when 'line-through'
        styles << :strikethrough
      else
        # -------------------------------------------------------------------
        # JBR69
        pvalue = attrs[:class]                      

        font_color = @theme[%(#{pvalue}_font_color)]            
        if (!font_color.nil? and font_color != '')
            if font_color.start_with? '#'
              # hexa color code
              font_color = font_color[1..-1] 
            end
            fragment[:color] = font_color            
        end

        font_family = @theme[%(#{pvalue}_font_family)]
        if (!font_family.nil? and font_family != '')
            fragment[:font] = font_family
        end  

        font_size = @theme[%(#{pvalue}_font_size)]
        if (!font_size.nil? and font_size != '')
            fragment[:size] = font_size
        end

        font_style = @theme[%(#{pvalue}_font_style)]
        if (!font_style.nil? and font_style != '')
            if font_style.include? 'italic'
              styles << :italic
            end
            if font_style.include? 'bold'            
              styles << :bold
            end
            if font_style.include? 'underline'
              styles << :underline
            end
            if font_style.include? 'subscript'
              styles << :subscript
            end
            if font_style.include? 'superscript'
              styles << :superscript
            end
            if font_style.include? 'strikethrough'
              styles << :strikethrough
            end          
        end 
        # ------------------------------------------------------------------- 
      end
    end if attrs.key?(:class)
    fragment.delete(:styles) if styles.empty?
    fragment
```

also replaces in this file, the variable `theme` by `@theme`

**Comment by namkhaib** [comment]
I'am _very_ new on Ruby (sorry)
When you say:
> @JBR69: also replaces in this file, the variable <code>theme</code> by <code>@theme</code>

it's _all_ the coincidences with <code>theme</code>, or a specific code?

**Comment by JBR69** [comment]
![image](https://user-images.githubusercontent.com/20417124/32365658-fba0fada-c07a-11e7-8af9-2c8a0ecef66e.png)

**Comment by namkhaib** [comment]
Thank you very much! This solves my ignorance :+1: :smile:

**Comment by mojavelinux** [comment]
This is implemented in 1.5.0.beta.2. Currently, you can set font properties (color, size, family, style). If more settings are needed, let's follow up with a specific issue.
